src/server.rs

```diff
@@ -619,28 +664,32 @@ impl Server {
     }
 
     fn execute_command(&mut self, id: RequestId, params: ExecuteCommandParams) -> Result<()> {
-        let db = self.engine.read();
-        match workspace_command::select(db, &params.command, params.arguments) {
-            Ok(command) => {
-                let client = self.client.clone();
-                self.pool.execute(move || {
-                    match command.run() {
-                        Ok(()) => {
-                            client
-                                .send_response(lsp_server::Response::new_ok(id, ()))
-                                .unwrap();
-                        }
-                        Err(why) => {
-                            client
-                                .send_error(id, ErrorCode::InternalError, why.to_string())
-                                .unwrap();
-                        }
-                    };
+        match params.command.as_str() {
+            "texlab.cleanAuxiliary" => {
+                self.run_errorable_with_db(id, move |db| {
+                    let opt = clean::CleanOptions::Auxiliary;
+                    clean::CleanCommand::new(db, opt, params.arguments)?.
```

Comment:
This closure has a subtle problem:

`db` is a snapshot of the database which prevents modifications while the snapshot is alive.
So when a query runs, the server needs to wait until the query is done before it can process notifications such as `textDocument/didChange`. That's why the server is essentially blocked while `latexmk` is running. Only the actual command should run on the thread pool but not the "preparation".

Reply:
When correcting this, I walked back on errorable forked commands. All commands that depend on the snapshot are at the moment `Option<T>` except for the `CleanCommand`. I did keep the refactored `run_erroable` function to maintain a shorter code for `exceute_command`.